presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();
+            ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
+            for (Expression conjunct : extractConjuncts(predicate)) {
+                List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);
+                if (correlation.stream().anyMatch(symbols::contains)) {
+                    joinExpressions.add(conjunct);
+                }
+                else {
+                    newPredicate.add(conjunct);
+                }
+            }
+
+            FilterNode newFilterNode = new FilterNode(
+                    idAllocator.getNextId(),
+                    filter.get().getSource(),
+                    ExpressionUtils.andOrTrue(newPredicate.build()));
+
+            if (!joinExpressions.build().isEmpty()) {
+                // filter condition has changed so Limit node no longer applies for EXISTS subquery
+                aggregationSource = searchFrom(aggregationSource)
+                        .where(LimitNode.class::isInstance)
+                        .skipOnlyWhen(ProjectNode.class::isInstance)
+                        .removeFirst();
+            }
+            aggregationSource = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .replaceAll(newFilterNode);
+
+            aggregationSource = ensureJoinSymbolsAreReturned(aggregationSource, joinExpressions.build());
+
+            return Optional.of(new RewrittenScalarAggregationSource(joinExpressions.build(), aggregationSource));
+        }
+
+        private PlanNode ensureJoinSymbolsAreReturned(PlanNode scalarAggregationSource, List<Expression> joinPredicate)
+        {
+            Set<Symbol> joinExpressionSymbols = DependencyExtractor.extractUnique(joinPredicate);
+            ExtendProjectionRewriter extendProjectionRewriter = new ExtendProjectionRewriter(
+                    idAllocator,
+                    joinExpressionSymbols);
+            return rewriteWith(extendProjectionRewriter, scalarAggregationSource);
+        }
+
+        private static boolean isSupportedPredicate(Expression predicate)
+        {
+            AtomicBoolean isSupported = new AtomicBoolean(true);
+            new DefaultTraversalVisitor<Void, AtomicBoolean>()
+            {
+                @Override
+                protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, AtomicBoolean context)
+                {
+                    if (node.getType() != LogicalBinaryExpression.Type.AND) {
```

Comment:
Possibly. Anyway I would prefer to merge conservative version first and then try to extend it to support more queries.

Supporting, more complicated predicates could be tricky.

Lets leave it as it is and then I will take care in separate PR as a separate issue (see https://github.com/prestodb/presto/issues/6154)